perllib/FixMyStreet/Cobrand/Bromley.pm

```diff
@@ -1256,24 +1258,40 @@ sub waste_reconcile_direct_debits {
                 $renew->insert;
                 $handled = 1;
             }
-        # If we're a cancel payment then find the initial subscription payment, also
-        # checking if we've already processed this payment. If we've not processed it
-        # create a cancel record using the original subscription as a basis.
-        # Note that cancel payments are marked as Processed and not Paid.
+        # There's two options with a cancel payment. If the user has cancelled it outside of
+        # WasteWorks then we need to find the original sub and generate a new cancel subscription
+        # report.
+        #
+        # If it's been cancelled inside WasteWorks then we'll have an unconfirmed cancel report
+        # which we need to confirm.
         } elsif ( $category eq 'Cancel' ) {
             next unless $payment->{Status} eq 'Processed';
             $rs = $rs->search({ category => { -in => ['Garden Subscription', 'Cancel Garden Subscription'] } });
-            my $p;
+            my ($p, $r);
             while ( my $cur = $rs->next ) {
                 my $sub_type = $cur->get_extra_field_value('Subscription_Type') || '';
                 if ( $sub_type eq $self->waste_subscription_types->{New} ) {
                     $p = $cur;
                 } elsif ( $cur->category eq 'Cancel Garden Subscription' ) {
+                    if ( $cur->state eq 'unconfirmed' ) {
+                        $r = $cur;
                     # already processed
-                    next RECORD if $cur->get_extra_metadata('dd_date') && $cur->get_extra_metadata('dd_date') eq $date;
+                    } elsif ( $cur->get_extra_metadata('dd_date') && $cur->get_extra_metadata('dd_date') eq $date) {
+                        next RECORD;
+                    }
                 }
             }
-            if ( $p ) {
+            if ( $r ) {
+                my $service = $self->waste_get_current_garden_sub( $r->get_extra_field_value('property_id') );
+                # if there's not a service then it's fine as it's already been cancelled
+                if ( $service ) {
+                    $r->set_extra_metadata('dd_date', $date);
+                    $r->state('confirmed');
```

Comment:
So if somehow a service has already been cancelled, but this $r report is still unconfirmed, the $payment will be processed each time it's fetched with no outcome? Should we mark the report as 'hidden' or something, if it's a report that doesn't need to be confirmed or do anything?